### PR TITLE
Introduce logic to extract and save API PB configuration from the general config file

### PR DESCRIPTION
**Pull request by strategio**
The `Hooks` claas will be implemented later in the API PB packages with a shared factory.

Please check the [reference for new allow items in the config file](https://onthegosystems.myjetbrains.com/youtrack/issue/wpmlcore-7560).

https://onthegosystems.myjetbrains.com/youtrack/issue/wpmlcore-7561

### CODE REVIEW

File: src/Config/Parser.php [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+	 *
+	 * @see WPML_Elementor_Translatable_Nodes::get_nodes_to_translate()
+	 *
+	 * @param array $allConfig
+	 *
+	 * @return array
+	 */
+	public function extract( array $allConfig ) {
+		$pbConfig = [];
+		$allWidgets = Obj::path( [ 'wpml-config', $this->configRoot, 'widget' ], $allConfig );
```
Comment (brucepearson) [inline src/Config/Parser.php:33]: @strategio I think you can use `pathOr( [], .... )` here to return an empty array and then remove the `if( $allWidgets )` check

File: src/Config/Parser.php [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+			$parsedConditions = [];
+
+			foreach ( $this->normalize( $rawConditions ) as $condition ) {
+				$parsedConditions[ Obj::pathOr( $this->defaultConditionKey, ['attr', 'key'], $condition ) ] = $condition['value'];
+			}
+		} else {
+			$parsedConditions = [ $this->defaultConditionKey => $widgetName ];
+		}
+
+		return $parsedConditions;
```
Comment (brucepearson) [inline src/Config/Parser.php:81]: @strategio You could try something like this. You'll need to make `normalize` public for this.

```php
$makePair = function( $condition ) { 
    return [Obj::pathOr( $this->defaultConditionKey, ['attr', 'key'], $condition ), $condition['value'] ] 
};
return Maybe::fromNullable( Obj::path( ['conditions', 'condition'], $widget ) )
   ->map( [ $this, 'normalize ] )
   ->map( Fns::map( $makePair ) )
   ->map( Lst::fromPairs() )
   ->getOrElse(  [ $this->defaultConditionKey => $widgetName ] );
```

File: src/Config/Parser.php [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+	}
+
+	/**
+	 * @param array $widget
+	 *
+	 * @return array
+	 */
+	private function parseFields( array $widget ) {
+		$rawFields    = Obj::path( ['fields', 'field'], $widget );
```
Comment (brucepearson) [inline src/Config/Parser.php:90]: @strategio Maybe use pathOr with a default of an empty array.

File: src/Config/Parser.php [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+		return $parsedFields;
+	}
+
+	/**
+	 * @param array $fieldsInItem
+	 *
+	 * @return array
+	 */
+	private function parseFieldsInItem( array $fieldsInItem ) {
+		$rawFieldsInItem    = Obj::prop( 'field', $fieldsInItem );
```
Comment (brucepearson) [inline src/Config/Parser.php:126]: @strategio Maybe propOr here

File: src/Config/Parser.php [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+	 */
+	private function parseIntegrationClasses( array $widget ) {
+		$rawClasses    = Obj::path( [ 'integration-classes', 'integration-class' ], $widget );
+		$parsedClasses = [];
+
+		if ( $rawClasses ) {
+			$parsedClasses = Lst::pluck( 'value', $this->normalize( $rawClasses ) );
+		}
+
+		return $parsedClasses;
```
Comment (brucepearson) [inline src/Config/Parser.php:158]: @strategio 

```php
return Obj::pathOr( [] [ 'integration-classes', 'integration-class' ], $widget )
   ->map( [ $this, 'normalize' ] )
   ->map( Lst::pluck( 'value' );
```

**Review by brucepearson — CHANGES_REQUESTED** [review]
@strategio The code looks good. I just made a few suggestions you might want to make.

File: tests/phpunit/tests/Config/TestStorage.php [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+	const PB_KEY = 'the-pb-key';
+
+	/**
+	 * @test
+	 */
+	public function itShouldGet() {
+		$pbConfig = [ 'some PB config' ];
+
+		$optionManager = $this->getOptionManager();
```
Comment (brucepearson) [inline tests/phpunit/tests/Config/TestStorage.php:21]: @strategio You can use the `OptionMock` trait for testing and then you don't need to mock OptionManager. You can just  use the Store get and set function to test.

```
$subject = $this->getSubject( new OptionManager() );
$subject->update( $pbConfig );

$this->assertEquals( $pbConfig, $subject->get() );
```

Comment (strategio) [reply tests/phpunit/tests/Config/TestStorage.php:21]: @brucepearson, it's not so simple becaue [OptionManager](https://git.onthegosystems.com/wpml/sitepress-multilingual-cms/-/blob/develop/classes/options/class-wpml-option-manager.php) class is in Core.

File: src/Config/Parser.php (hunk 1 above)
Comment (strategio) [reply src/Config/Parser.php:33]: Done!

File: src/Config/Parser.php (hunk 2 above)
Comment (strategio) [reply src/Config/Parser.php:81]: Excellent @brucepearson , done!

File: src/Config/Parser.php (hunk 3 above)
Comment (strategio) [reply src/Config/Parser.php:90]: Done

File: src/Config/Parser.php (hunk 4 above)
Comment (strategio) [reply src/Config/Parser.php:126]: Done

File: src/Config/Parser.php (hunk 5 above)
Comment (strategio) [reply src/Config/Parser.php:158]: Adjusted and added.

**Review by strategio — COMMENTED** [review]
Thanks @brucepearson! I was also planning to improve the code yesterday but I believe it was a bit late...

**Comment by strategio** [comment]
Ooops, I forgot to push the last changes. This is ready now!

File: tests/phpunit/tests/Config/TestStorage.php (hunk 6 above)
Comment (brucepearson) [reply tests/phpunit/tests/Config/TestStorage.php:21]: @strategio I forgot. I did have a plan to move this in WP library. I might do it sometime soon to make these things easier.

**Review by brucepearson — APPROVED** [review]
@strategio Looks good. +1

**Review by dgwatkins — APPROVED** [review]
:+1: @strategio

**Pull request merged**